File: probs/1306.cpp
#include <queue>
#include <vector>
using namespace std;

class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        queue<pair<int, int> > q;
        q.push({ start, 0 });

        vector<bool> visited(arr.size(), false);

        while (!q.empty()) {
            pair<int, int> p = q.front();
            q.pop();

            int pos = p.first;
            int count = p.second;

            visited[pos] = true;

            if (arr[pos] == 0) {
                return true;
            }

            int a = pos + arr[pos];
            int b = pos - arr[pos];

            if (a >= 0 && a < arr.size() && !visited[a]) {
                q.push({ a, count + 1 });
            }

            if (b >= 0 && b < arr.size() && !visited[b]) {
                q.push({ b, count + 1 });
            }
        }

        return false;
    }
};

File: go/1306.go
package main

func canReach(arr []int, start int) bool {
	visited := make([]bool, len(arr))
	visited[start] = true

	q := []int{}
	q = append(q, start)

	for len(q) > 0 {
		curr := q[0]
		q = q[1:]

		if arr[curr] == 0 {
			return true
		}

		left := curr - arr[curr]
		if left >= 0 && !visited[left] {
			visited[left] = true
			q = append(q, left)
		}

		right := curr + arr[curr]
		if right < len(arr) && !visited[right] {
			visited[right] = true
			q = append(q, right)
		}
	}

	return false
}
